tests/integrationtest/l4_crd_test.go

```diff
@@ -0,0 +1,703 @@
+	"google.golang.org/protobuf/proto"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func validateCRDValues(t *testing.T, g *gomega.GomegaWithT, expectedValues interface{}, actualValues ...interface{}) {
+	valuesfromCRD := structs.Map(expectedValues)
+	valuesFromGraphLayer := make(map[string]interface{})
+	for _, actualValue := range actualValues {
+		actualValueMap := structs.Map(actualValue)
+		maps.Copy(valuesFromGraphLayer, actualValueMap)
+	}
+
+	for field, valuefromCRD := range valuesfromCRD {
+
+		t.Logf("Current Field: %s", field)
+		valueFromGraphLayer, ok := valuesFromGraphLayer[field]
+		if !ok {
+			continue
+		}
+
+		if strings.HasSuffix(field, "Ref") {
+			ref1 := valueFromGraphLayer.(*string)
+			ref2 := valuefromCRD.(*string)
+			g.Expect(ref1).NotTo(gomega.BeNil())
+			g.Expect(ref2).NotTo(gomega.BeNil())
+			g.Expect(*ref1).To(gomega.HaveSuffix(*ref2))
+		} else if strings.HasSuffix(field, "Refs") {
+			refs1 := valueFromGraphLayer.([]string)
+			refs2 := valuefromCRD.([]string)
+			g.Expect(len(refs1)).To(gomega.Equal(len(refs2)))
+			l := len(refs1)
+			for i := 0; i < l; i++ {
+				g.Expect(refs1[i]).To(gomega.HaveSuffix(refs2[i]))
+			}
+		} else if field == "AnalyticsPolicy" {
+			// All fields of AnalyticsPolicy are not used currently. Hence
+			// handling it separately.
+			actualAnalyticsPolicy := valueFromGraphLayer.(map[string]interface{})
+			expectedAnalyticsPolicy := valuefromCRD.(map[string]interface{})
+			// For backendProperties, the PoolAnalyticsPolicy only have one field EnableRealtimeMetrics
+			if _, ok := expectedAnalyticsPolicy["EnableRealtimeMetrics"]; ok {
+				g.Expect(expectedAnalyticsPolicy["EnableRealtimeMetrics"]).To(gomega.Equal(actualAnalyticsPolicy["EnableRealtimeMetrics"]))
+			} else {
+				g.Expect(expectedAnalyticsPolicy["FullClientLogs"]).To(gomega.Equal(actualAnalyticsPolicy["FullClientLogs"]))
+			}
+		} else {
+			g.Expect(utils.Stringify(valuefromCRD)).To(gomega.Equal(utils.Stringify(valueFromGraphLayer)))
+		}
+	}
+}
+
+func TestCreateDeleteL4Rule(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	L4RuleName := "test-l4rule"
+	ports := []int{8080}
+
+	SetUpTestForSvcLB(t)
+
+	g.Eventually(func() bool {
+		found, _ := objects.SharedAviGraphLister().Get(SINGLEPORTMODEL)
+		return found
+	}, 30*time.Second).Should(gomega.Equal(true))
+	_, aviModel := objects.SharedAviGraphLister().Get(SINGLEPORTMODEL)
+	nodes := aviModel.(*avinodes.AviObjectGraph).GetAviVS()
+	g.Expect(nodes).To(gomega.HaveLen(1))
+	g.Expect(nodes[0].Name).To(gomega.Equal(fmt.Sprintf("cluster--%s-%s", NAMESPACE, SINGLEPORTSVC)))
+	g.Expect(nodes[0].Tenant).To(gomega.Equal(AVINAMESPACE))
+	g.Expect(nodes[0].PortProto[0].Port).To(gomega.Equal(int32(8080)))
+
+	// Check for the pools
+	g.Expect(nodes[0].PoolRefs).To(gomega.HaveLen(1))
+	address := "1.1.1.1"
+	g.Expect(nodes[0].PoolRefs[0].Servers[0].Ip.Addr).To(gomega.Equal(&address))
```

Comment:
@swathinsankaran : What is reason of checking pool server IP address?

Reply:
We are just making sure the pool properties are correct. You're right. It is not be relevant from an L4 CRD perspective.